### ISSUE TITLE
Handling responses with types dependent on status codes

### ISSUE THREAD

**Issue by timbod7** [purescript-contrib/purescript-affjax#34]
I am making requests on a service that returns status code 200 and  a body containing json data on success, but returns status code 401 and a plain error message when unauthorised.

How should this be handled? The precise type of the AffjaxResponse can't really be determined until the status code is examined.

**Comment by jdegoes** [comment]
That's a bad API. If you request JSON response, then the error should be encoded in JSON.

That said, we have to deal with APIs. :smile: 

For a short-term workaround, I'd suggest creating a `UnknownFormat` type which is basically just a `String`, which you can then de-serialize to JSON or interpret as text depending on the response code. I'll try to think of a better to handle this in the API (maybe by having such a type).

**Comment by timbod7** [comment]
I'm just using Strings for now, and doing the deserialisation outside the
response. It works ok, but it is a shame to not be using the "baked in"
decoding.

**Comment by lattenwald** [comment]
I think my problems belongs to this ticket.

Reddit API correctly returns 403 when my authentication header is missing or is wrong. How can this be handled with affjax? I want an obvious thing — to get status code as an Integer so I could pattern-match on it or check equality with it.

**Comment by garyb** [comment]
The status code is accessible via the `status` property of the response - although if the `content-type` of the response is typed differently than what was requested, then yes this ticket is relevant, as the request will fail entirely.

If this is the case, then the workaround is to expect an `Foreign` result, and then handle the parsing of the content yourself, choosing the appropriate parser for the status code returned.

**Comment by lattenwald** [comment]
I didn't elaborate.
I get an Error. When I use `Aff.attempt`, I get `Left` value with `Error` with a string inside, can I get status from there?

**Comment by garyb** [comment]
No you can't - that's what I meant by the "request failing entirely" (unless the string contains the error code or something like that and you parse it out of there, but that's not really the point!).

**Comment by spicydonuts** [comment]
I have an edge case that might fit into this.  The API speaks Json but may return a `201` status with a `Content-Length: 0` response header.  `null` is valid in `JSON.parse`, but `undefined` and `''` are not.  At first glance this made sense.. no body was returned so it's not Json!  _But_ `null` is valid Json.. and it might be a much cleaner way to encode this edge case.  Or maybe the response type should be `Maybe Json`?  Looking at the code a bit, I think I'll have the same issue using `Foreign`, because it uses `JSONResponse` as well, which is what causes it to run the response through `parseJSON`.. maybe it could convert empty strings to `null` and send everything else through `parseJSON`.

Or maybe I should just be doing it all manually with `String` 😅

**Comment by garyb** [comment]
@spicydonuts I don't think we call `parseJSON` at all actually - it `unsafeCoerce`s values in the `Respondable` instances for `Foreign` and `Json`, so I'm not too sure what's going on your case.

**Comment by spicydonuts** [comment]
This case: https://github.com/slamdata/purescript-affjax/blob/master/src/Network/HTTP/Affjax.purs#L283

**Comment by garyb** [comment]
Well that's embarrassing :see_no_evil:

**Comment by spicydonuts** [comment]
Haha, no worries.  I currently have a `MaybeBody a` newtype around `Maybe a` which still sets the JSON content type headers but uses `StringResponse`.  It returns `Nothing` if the `Foreign` handed to `fromResponse` is just an empty string.  It's one more failure case for the calling code to worry about, but it's easy to push `Nothing`s back up into the failure path with a `throwError`.  I don't think it's ideal though.. not sure what would be better for the default json-like instances.

**Comment by garyb** [comment]
I'm not sure what the right answer is here either. I guess the question is: is it ever going to be useful to distinguish between an empty response and an actual `null` value for the response. I'd lean towards "probably not". Or it would be uncommon at least.

With that, perhaps we should add some handling for empty results in the `JSONResponse` case. For people who do need that distinction, it's still possible, since the `String` result type could be used and then the JSON parsing could be done "manually".

**Comment by jdegoes** [comment]
Null is not the same as an empty response. A REST API that is returning `application/json` for its content type but which has an empty body is broken. Different applications may want to pretend this makes sense, but probably in different ways. As per spec, a `null` value may not even occur at the top-level position, so returning `null` there would be broken in other ways.

**Comment by spicydonuts** [comment]
That's true if you're coding exactly to the http spec, but if it doesn't affect the normal behavior at all it seems odd to have all the convenience break down because an api misbehaved slightly.

But that's a separate issue because the api isn't actually misbehaving.. it sends 200 and application/json OR 204 and no content-type and no body.  And Affjax doesn't use the response content-type when deciding whether to decode.. it uses the type built into the code, which _does_ make sense, but it'd be nice if `Maybe a` was easier.  Also, the parsing error is a slightly confusing "unexpected token 'u'" (trying to parse `undefined`) instead of something like "you said there'd be a body and there wasn't" or "content-type didn't match", both of which should happen before unsuccessful json parsing if coding exactly to the spec is the goal.

**Comment by garyb** [comment]
The type "built into the code" should be what is returned by the API, as it sends the appropriate `Accept` header for that content type - so that the API ignores that also means it's wrong, strictly speaking.

**Comment by spicydonuts** [comment]
There's a content-type for no body and the client should know in advance what the server will say, or no api should ever send differing response codes?

**Comment by spicydonuts** [comment]
Accept doesn't control whether the server sends a body in the response.  It only indicates to the server what formats the client prefers to receive data _if_ it does receive some.

**Comment by centromere** [comment]
Hi @garyb. I'm adding my comment to this thread because I believe it's relevant here.

I'm working with an API where the `/register` endpoint accepts a request body of Content-Type `application/json` and returns a 204 "No Content" status upon success (along with a zero length response), and upon failure a 4xx status and a `{ "error": "The foo provided is invalid" }` response body. This behavior seems impossible to encode in Affjax. Is the endpoint violating the HTTP spec? What content types should it `Accept`?

**Comment by garyb** [comment]
I think if the response `Content-Type` is `application/json` then the 204 should probably also return `{}` rather than a zero-length response. I could be wrong though, in which case maybe we need to add a special case for zero-length JSON responses?

This behaviour is impossible for Affjax to deal with currently though, yeah - I'd suggest just using a string reponse type, and then switching on the status code to "manually" `parseJson` as necessary instead.

**Comment by centromere** [comment]
[It's invalid](https://tools.ietf.org/html/rfc2616#section-10.2.5) to return `{}` with a 204:

> The 204 response MUST NOT include a message-body, and thus is always
> terminated by the first empty line after the header fields.

Why should the `Content-Type` of the request necessarily match the `Content-Type` of the response?

**Comment by garyb** [comment]
Ah right. 

> Why should the `Content-Type` of the request necessarily match the `Content-Type` of the response?

I was saying if the response `Content-Type` is `application/json` and it's empty I think that's probably incorrect - not the request. Even if it's not, this won't work out with affjax as it stands, as it will either fail due to the response `Content-Type` not being `application/json` or due to it not parsing `""` for the `Respondable` instance for `Json`.

**Comment by centromere** [comment]
Ah, but what if the `Content-Type` of the 204 response is unspecified? It's not mandatory.

Moreover, what should the `Accept` header be set to in the request?

**Comment by garyb** [comment]
What I mean is, if you use a `Respondable` of type `Json`, it will fail _unless_ `Content-Type` is `application/json`, etc. Hence why I suggest using a string and then decoding manually when the response is determined.

**Comment by spicydonuts** [comment]
When I ran into this I ended up making a custom decoder around Maybe.  `204`s become `Nothing`.  It was kind of obnoxious to write though.  I'd rather be forced to deal with the Maybe all the time if it makes it easy to cover potential `204`s.

Specifically,
```purs
type AffjaxResponse a =
  { status :: StatusCode
  , headers :: Array ResponseHeader
  , response :: a
  }
```
would become
```purs
type AffjaxResponse a =
  { status :: StatusCode
  , headers :: Array ResponseHeader
  , response :: Maybe a
  }
```
because it's always possible for a server to send an empty body.

**Comment by thomashoneyman** [comment]
This is no longer relevant now that there is not a `Requestable` or `Respondable` type, so I'm going to close the issue. Thanks!
